fuels-rs/src/contract.rs

```diff
@@ -353,24 +361,26 @@ where
         .await
         .unwrap();
 
-        let returned_value = match Self::get_receipt_value(&receipts) {
-            Some(val) => val.to_be_bytes(),
-            None => [0u8; 8],
-        };
-
-        let mut decoder = ABIDecoder::new();
-
-        let decoded = decoder.decode(&self.output_params, &returned_value)?;
-
-        Ok(D::from_tokens(decoded)?)
-    }
-
-    fn get_receipt_value(receipts: &[Receipt]) -> Option<u64> {
+        let mut encoded_value = None;
+        let mut contract_receipts = Vec::new();
         for receipt in receipts {
             if receipt.val().is_some() {
-                return receipt.val();
+                if encoded_value.is_some() {
+                    contract_receipts.push(receipt)
+                } else {
+                    encoded_value = receipt.val();
+                }
             }
         }
-        None
+        let encoded_value = match encoded_value {
+            Some(val) => val.to_be_bytes(),
+            None => [0u8; 8],
+        };
```

Comment:
This gets the job done, but we can make this much simpler and efficient: 

```Rust
pub async fn call(self) -> Result<CallResponse<D>, Error> {
    let receipts = Contract::call(
        // ...
    )
    .await
    .unwrap();

  // No need to create a whole new `contract_receipts` (we already have the receipts returned from the call!)
  // No need for a `for-loop` with two nested `if`s, better for readability.
  // So here we're basically doing the `receipt.val()` look-up and parsing in one go. 
  let encoded_value = match receipts.iter().find(|&receipt| receipt.val().is_some()) {
      Some(r) => r.val().unwrap().to_be_bytes(),
      None => [0u8; 8],
  };
  
  let mut decoder = ABIDecoder::new();
  
  let decoded_value = decoder.decode(&self.output_params, &encoded_value)?;
  
  Ok(CallResponse {
      value: D::from_tokens(decoded_value)?,
      receipts, // And here we pass the `receipts` that returned from the call itself
  })
}
```
Much shorter, eh? Let me know what you think! Also all the comments I placed here are just for explanation, they wouldn't go in the actual code! 😅

Reply:
But shouldn't we call `.val()` on the `receipts` other than just the *one* value? And also we are returning the `value` in receipts (so twice?).

Reply:
Not sure I understand what you mean.

> But shouldn't we call .val() on the receipts other than just the one value?

Should we? What we want: the value in the receipt type that contains the value returned by the method we call, so only one: the `Return` receipt that, which is the one that contains `.val() == Some` (https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/abi.md#return-receipt)

> And also we are returning the value in receipts (so twice?)

Not sure what you mean by this.

Reply:
Ok go it!
> Not sure what you mean by this.
I just mean that the `encoded_value` (the one from which we derive `value`) is also inside the `receipts` vector so it's one thing we could remove from the receipts vector. We are sending the same information twice (once decoded as `value`, the other one still encoded, the first non-None item in the `receipts` vector). I'm not sur I make sense :sweat_smile:

Reply:
Oh, I see what you mean! Yeah, we could remove that receipt out of the `receipts`, since `value` is already being sent to the user. I guess we can declare `receipts` as `mut` and remove from it once we decode the `Return` receipt.